simulations/validator_client.py

```diff
@@ -0,0 +1,32 @@
+from utils.clock import Clock
+
+
+class ValidatorClient(Clock):
+    """Validator client software that makes decisions about when
+    to make and send new messages"""
+    def __init__(self, validator, network):
+        super().__init__()
+        self.validator = validator
+        self.network = network
+
+    def should_make_new_message(self):
+        pass
+
+    def make_new_message(self):
+        if self.should_make_new_message():
+            return self.validator.make_new_message()
+
+    def retrieve_messages(self):
+        messages = self.network.receive_all_available(self.validator)
+        self.validator.receive_messages(set(messages))
+        return messages
+
+    def propagate_message(self, message):
+        """ could change this out to send to only peers """
+        self.network.send_to_all(message)
+
+    def make_and_propagate_message(self):
```

Comment:
This seems like a misleading function name. Maybe add `try` in there somewhere or something (not sure, just something to signify that a new message might not be made...)

Reply:
word

Reply:
the call to `should` shouldn't be in either of these `make_message` methods. Poor design. The caller of `make_message` should do the check on `should` and then proceed accordingly. This will allow how to override if necessary (not sure when he would want to), but it more importantly will allow the validator_client to do something else if it's not time to make a message.